---
title: "Data QC report"
subtitle: "DataQC v`r params$dataqc_version`"
date: "`r format(Sys.time(), '%d/%m/%y')`"
output: 
  html_document:
    code_folding: hide
    toc: true
params:
    dataqc_version: "unknown"
    dataset_name: "unknown"
    platform: "unknown"
    N: 100
    S: 0.4
    SD: 3
    SD_exp: 4
    ContRatio: "fixed"
    Cont: 30
    ContSlope: 45
---

# Dataset

This report is for cohort: **`r params$dataset_name`**.

Expression platform for this cohort is: **`r params$platform`**.

Post-QC sample size for this cohort is: **`r library(data.table); exp <- fread(params$N); nrow(exp)`**

**Settings used for sample outlier detection:**

LOF S-threhold for genetic outliers: `r params$S`

Threshold of SD difference from mean genetic PC1 and PC2 is: `r params$SD`

Threshold of SD difference from mean gene expression PC1 and PC2 is: `r params$SD_exp`

Scales of sex-chromosome expression for expression-based sex determination: `r params$ContRatio`

Angle of contamination area in sex chromosome gene expression plot: `r params$Cont`

Slope in sex chromosome gene expression plot: `r params$ContSlope`

## Covariates

```{r, message = FALSE, warning = FALSE, echo = FALSE}
library(data.table)
library(knitr)
library(ggplot2)
library(stringr)

cov <- fread("CovariatePCs.txt", keepLeadingZeros = TRUE, colClasses = list(character = c(1)))
nr_add_cov <- 0
```

`DataQC` pipeline has prepared `r ncol(cov) - 1` covariates: 10 first genetic PCs`r if (ncol(cov) > 111){ab <- ','}else{ab <- ' and'}; ab` 100 first expression PCs`r if (ncol(cov)> 111){outp <- paste0(' and ', ncol(cov) - 111, ' additional covariate(s).')} else {outp <- '.'}; outp`

`r if (ncol(cov) > 111){outp <- paste0('The names of additional covariates are: ', paste(colnames(cov)[-c(1:111)], collapse = ", "), '.'); outp}`

```{r, message = FALSE, eval = ncol(cov) > 111, warning = FALSE, echo = FALSE}
cov_specific <- melt(cov, id.vars = "SampleID")
colnames(cov_specific) <- c("SampleID", "covariate", "value")
cov_specific <- cov_specific[!(str_detect(cov_specific$covariate, "GenPC") | str_detect(cov_specific$covariate, "ExpPC")), ]

cov_specific$covariate <- as.character(cov_specific$covariate)
warning("COV SPECIFIC!!!")
warning(cov_specific)
nr_add_cov <- length(unique(cov_specific$covariate))
```

```{r, fig.height = 7, fig.width = 7, fig.align = "center", message = FALSE, eval = (nr_add_cov == 1 | nr_add_cov == 7 | nr_add_cov == 8 | nr_add_cov == 9), warning = FALSE, echo = FALSE}
ggplot(cov_specific, aes(x = value)) + geom_histogram(colour = "white") + facet_wrap(~ covariate, scales = "free", ncol = 3) + theme_bw() + xlab("Covariate value")
```

```{r, fig.height = 3.5, fig.width = 7, fig.align = "center", message = FALSE, eval = (nr_add_cov == 2 | nr_add_cov == 3 | nr_add_cov == 4 | nr_add_cov == 5 | nr_add_cov == 6), warning = FALSE, echo = FALSE}
ggplot(cov_specific, aes(x = value)) + geom_histogram(colour = "white") + facet_wrap(~ covariate, scales = "free", ncol = 3) + theme_bw() + xlab("Covariate value")
```

```{r, fig.height = 15, fig.width = 7, fig.align = "center", message = FALSE, eval = (nr_add_cov > 9 & nr_add_cov < 21), warning = FALSE, echo = FALSE}
ggplot(cov_specific, aes(x = value)) + geom_histogram(colour = "white") + facet_wrap(~ covariate, scales = "free", ncol = 3) + theme_bw() + xlab("Covariate value")
```

```{r, fig.height = 15, fig.width = 7, fig.align = "center", message = FALSE, eval = (nr_add_cov > 20), warning = FALSE, echo = FALSE}
ggplot(cov_specific, aes(x = value)) + geom_histogram(colour = "white") + facet_wrap(~ covariate, scales = "free", ncol = 4) + theme_bw() + xlab("Covariate value")
```

`r if (nr_add_cov == 0) {"\\begin{comment}"}`
**Fig 1. Histograms for additional covariates.**
`r if (nr_add_cov == 0) {"\\end{comment}"}`

`r if (nr_add_cov > 0) {"\\begin{comment}"}`
**Fig 1. Histograms for additional covariates.** Not available, there are no additional covariates prepared on top of genotype PCs and expression PCs.
`r if (nr_add_cov > 0) {"\\end{comment}"}`

## Genotype QC report

### Overview of the QC steps

```{r, message = FALSE, warning = FALSE, echo = FALSE, results = 'asis'}

wgs_table <- fread("wgs_qc_table_combined.txt", header = TRUE, data.table = FALSE)
#wgs_settings <- fread("wgs_qc_settings_combined.txt", header = TRUE, data.table = FALSE)
overview_table <- fread("outputfolder_gen/gen_data_summary/summary_table.txt", data.table = FALSE)

# Check what the number of SNPs is that went into WGS specific QC
if (nrow(wgs_table != 0)) {

  #wgs_table$Autosome <- ifelse(wgs_table$Chromosome %in% paste0("chr", c(1:22)), "Autosomes", "X chromosome")
  #aggregated <- aggregate(wgs_table[columns_to_aggregate], by=list(wgs_table$Autosome), FUN=sum)

  #rownames(aggregated) <- aggregated$Group.1

  #wgs_overview_table <- t(aggregated[columns_to_aggregate])
  #rownames(wgs_overview_table) <- column_labels[rownames(wgs_overview_table)]
  #knitr::kable(wgs_overview_table)

  #new_raw_row <- overview_table[overview_table["Stage"] == "Raw file"]
  #new_raw_row["Nr. of SNPs"] <- sum(wgs_table[,"PASSQC"])
  overview_table[overview_table["Stage"] == "Raw file", "Stage"] <- "Passed WGS specific SNP QC"
  #overview_table <- rbind(data.frame(new_raw_row, overview_table))
}
# List the set of variants

knitr::kable(overview_table)
```

### Allele frequency concordance with reference panel

In the following plot you will see the comparison of allele frequencies between input genotypes and 1000G phase 3 reference panel. This lookup is done on SNP IDs, therefore many SNPs will be excluded. However, even with the subset of SNPs, we can check if the allele frequencies generally correlate between your cohort and reference dataset. We should see that a large majority of the variants are on the 0-1 diagonal and there is a strong correlation between the allele frequencies of two datasets.

However:

- If you see the majority of the dots on 1-0 diagonal (you see strong **negative** correlation), this suggest that your input genotypes have incorrect encoding of the alleles for all SNPs.
- If you see inverted "V" shape, this suggests that there are some subset of SNPs which have alleles incorrectly encoded.

In both of the cases you should check if the input `.bed/.bim/.fam` files are correct and that there are no errors made in generating and/or processing those. If you do not pinpoint any issues with the input, please contact the lead analyst(s) for advice.

```{r, message = FALSE, warning = FALSE, echo = FALSE, fig.width = 6, fig.height = 6, fig.align = "center"}
target <- fread("target.afreq.gz", colClasses=c("character", "integer", "character", "character", "character", "numeric", "integer"))
ref <- fread("1000Gref.afreq.gz", colClasses=c("character", "integer", "character", "character", "character", "numeric", "integer"))

target <- target[!is.na(target$ALT_FREQS), ]
target <- target[, c(1:6), with = FALSE]
colnames(target)[3:6] <- paste0("target_", colnames(target)[3:6])

ref <- ref[, c(1:6), with = FALSE]
colnames(ref)[3:6] <- paste0("ref_", colnames(ref)[3:6])

comb <- merge(ref, target, by = c("#CHROM", "POS"))

comb2 <- comb[!(comb$ref_REF != comb$target_REF & comb$ref_REF != comb$target_ALT)]

# Remove ambiguous SNPs
comb_f <- comb2[!((comb2$ref_REF == "A" & comb2$ref_ALT == "T") | 
(comb2$ref_REF == "T" & comb2$ref_ALT == "A") |
(comb2$ref_REF == "C" & comb2$ref_ALT == "G") | 
(comb2$ref_REF == "G" & comb2$ref_ALT == "C")), ]

comb_f[comb_f$ref_REF == comb_f$target_ALT & comb_f$ref_ALT == comb_f$target_REF, ]$target_ALT_FREQS <- 1 - comb_f[comb_f$ref_REF == comb_f$target_ALT & comb_f$ref_ALT == comb_f$target_REF, ]$target_ALT_FREQS

ggplot(comb_f, aes(x = target_ALT_FREQS, y = ref_ALT_FREQS)) + geom_point(alpha = 0.2, size = 0.3) + 
theme_bw() + xlab("AF (input data)") + ylab("AF (1000G reference)")

res_cor <- cor(comb_f$ref_ALT_FREQS, comb_f$target_ALT_FREQS, use = "complete.obs")

```

**Fig 2. Comparison of allele frequencies between input genotypes and 1000G phase 3 reference.** Comparison is restricted to up to ~1.7M HapMap3 or UK Biobank SNPs as provided by `bigsnpr` R package. Ambiguous variants (A/T, C/G SNPs) and variants whose alleles seem to be from opposite strands in two datasets were removed from this analysis. 

Out of `r nrow(target)` variants in the input genotype data, `r nrow(comb)` overlap with `r nrow(ref)` variants in the reference data. Out of those, `r nrow(comb[(comb$ref_REF != comb$target_REF & comb$ref_REF != comb$target_ALT)])` are removed from this lookup due to difference of strand between reference and input and `r nrow(comb2[((comb2$ref_REF == "A" & comb2$ref_ALT == "T") | (comb2$ref_REF == "T" & comb2$ref_ALT == "A") | (comb2$ref_REF == "C" & comb2$ref_ALT == "G") | (comb2$ref_REF == "G" & comb2$ref_ALT == "C")), ])` are removed because these are ambiguous variants. This leaves `r nrow(comb_f)` variants for calculating correlations.

The Pearson R of allele frequencies is **`r res_cor`**, which suggests `r if(res_cor > 0 & res_cor < 0.5){quality_res <- "that there is something clearly wrong with your input genotypes. You should double-check your input data."} else if (res_cor > 0.5 & res_cor < 0.8){quality_res <- "that you might have something wrong with your input genotypes. Allele frequencies between two datasets correlate, however we would expect higher correlation. You should double-check your input data and contact with lead analysts for advice."} else if (res_cor > 0.8){quality_res <- "that your input data does not have major problems with allele encoding."} else if (res_cor < 0){quality_res <- "you might have systematic issues with allele encoding in your input genotypes. You should double-check your input data."}; quality_res`

### Sex check

`r if (!file.exists("outputfolder_gen/gen_plots/SexCheck.png")) {"\\begin{comment}"}`
Next plot summarises the genetically determined sex of the input samples. Samples which have unclear genetic sex (X chr heterozygosity >0.2 and <0.8) and hence might be contaminated, are removed. If reported sex is indicated in the input `.fam` file, this is also compared with genetic sex and problematic samples (potential mixups) are automatically excluded.

`r nrow(fread("outputfolder_gen/gen_data_QCd/SexCheckFailed.txt", keepLeadingZeros = TRUE))` samples failed the sex check and were removed.

You can investigate the file `[your output folder]/outputfolder_gen/gen_data_QCd/SexCheckFailed.txt` to see which samples failed the sex check.

![**Fig 3. Distribution of F-values representing X chromosome heterozygosity.** Genetically determined sex thresholds are shown as vertical lines. Samples with the F-value between those two lines are potentially contaminated samples and are automatically excluded.](outputfolder_gen/gen_plots/SexCheck.png)

`r if (!file.exists("outputfolder_gen/gen_plots/SexCheck.png")) {"\\end{comment}"}`

`r if (file.exists("outputfolder_gen/gen_plots/SexCheck.png")) {"\\begin{comment}"}`
Warning! This step was skipped since no X chromosome was found in the genotype data.
Use raw genotype data that includes the X chromosome if possible.

**Fig 3. Distribution of F-values representing X chromosome heterozygosity.** Not available!

`r if (file.exists("outputfolder_gen/gen_plots/SexCheck.png")) {"\\end{comment}"}`

### Excess heterozygosity

Next plot summarises the heterozygosity of the input samples. Samples showing excess/depleted heterozygosity rate (mean+/-3SD) have potential issues with contamination or inbreeding, and are automatically excluded.

![**Fig 4. Distribution of heterozygosity rate.** Thresholds for declaring excess/depleted heterozygosity for sample (mean+/-3SD) are shown as vertical lines. ](outputfolder_gen/gen_plots/HetCheck.png)  

### Ancestry outliers

#### Genotype data projected into 1000G superpopulations

Here we project your genotype samples into 1000G p3v5 superpopulations to check if all the samples come from one ancestry group. 

1. If you observe a few samples which are different ancestry compared to the majority of samples (e.g. majority of the samples are of EUR ancestry and there are few samples from AFR or SAS), then please remove ancestry outliers and re-run the `DataQC` pipeline without those samples.
2. If you observe many (>100) samples which are different ancestry compared to the majority of samples (your dataset consists of individuals from several ancestries) then please split your genotype data based on ancestry group and rerun this and following pipelines on those datasets separately.

The files `AssignedPopulations.txt` and `1000G_PC_projections.txt` have been written to the output folder under subfolder `gen_data_summary`. These can be used to extract the outlier samples based on their PC projections into 1000G superpopulations. Please use optional `--ExclusionList` and specify the file with the list of samples you want to withhold from this dataset.

![**Fig 5. Genotype samples projected into 1000G superpopulations.** First two plots show only target dataset in the 1000G PC space. Rest of the plots visualise target dataset, as well as samples from 1000G superpopulations by plotting first 10 genotype PCs.](outputfolder_gen/gen_plots/SamplesPCsProjectedTo1000G.png)  

#### Assign superpopulation for each sample

Based on first 3 genotype PCs, here we assign the most likely population to each sample and outline those samples which are clearly different. Again, all the samples in the dataset should be the most similar to only one ancestry and samples resembling other ancestries should be removed or treated as a separate dataset (in case of N>100).

```{r, fig.height = 7, fig.width = 7, fig.align = "center", echo = FALSE, message = FALSE, warning = FALSE}
library(dplyr)
pops <- fread("outputfolder_gen/gen_data_summary/PopAssignResults.txt", keepLeadingZeros = TRUE)

pops <- melt(pops, id.vars = "abi")
colnames(pops) <- c("Sample", "Superpopulation", "dist") 

pops <- pops %>% 
group_by(Sample) %>% 
arrange(dist, .by_group = TRUE) %>% 
mutate(SamplePopAssign = head(Superpopulation, 1)) %>%
mutate(Difference = dist[2]/dist[1]) %>%
mutate(Confidence = case_when(Difference > 2 ~ "confident", Difference <= 2 ~ "not confident"))

res <- unique(pops[, c(1, 4, 5, 6)])

# Visualise
ggplot(pops, aes(x = Superpopulation, y = dist, fill = Superpopulation)) + 
geom_point() + 
geom_violin(draw_quantiles = 0.5) + 
theme_bw() + 
scale_fill_manual(values = c("Target" = "black", 
"EUR" = "blue", 
"EAS" = "goldenrod", 
"AMR" = "lightgrey", 
"SAS" = "orange", 
"AFR" = "red"))

fwrite(res, "outputfolder_gen/gen_data_summary/AssignedPopulations.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

**Fig 6. Euclidean distance between target samples and samples from each 1000G superpopulation.** 

`r nrow(res[res$SamplePopAssign == "EUR",])` samples in the data are most similar to EUR superpopulation.

`r nrow(res[res$SamplePopAssign == "AMR",])` samples in the data are most similar to AMR superpopulation.

`r nrow(res[res$SamplePopAssign == "EAS",])` samples in the data are most similar to EAS superpopulation.

`r nrow(res[res$SamplePopAssign == "SAS",])` samples in the data are most similar to SAS superpopulation.

`r nrow(res[res$SamplePopAssign == "AFR",])` samples in the data are most similar to AFR superpopulation.

`r res %>% filter(Confidence != "confident") %>% nrow()` samples assignment to the data is not very confident (<2 fold difference with second most similar superpopulation).

Again, please follow the following instructions:

1. If you observe a few samples which are different ancestry compared to the majority of samples (e.g. majority of the samples are of EUR ancestry and there are few samples from AFR or SAS), then remove ancestry outliers and re-run the `DataQC` pipeline without those samples.
2. If you observe many (>100) samples which are different ancestry compared to the majority of samples (your dataset consists of individuals from several ancestries) then please split your genotype data, based on ancestry and rerun this and following pipelines on those datasets separately. This assumes that you know the ancestry information for your genotype data.

The files `AssignedPopulations.txt` and `1000G_PC_projections.txt` have been written to the output folder under folder `gen_data_summary`. These can be used to extract the outlier samples based on their PC projections into 1000G superpopulations. Please use optional `--ExclusionList` and specify the file with the list of samples you want to withhold from this dataset. If the large majority of your cohort is from specific superpopulation (e.g. EUR) but you observe a small number of samples which are not confidently assigned to this superpopulation (<2 fold difference with second most similar superpopulation), you can remove those too.

### Individual genetic outliers in the target population

Here we conduct PCA on the genotypes of eQTL samples only and investigate if there are any individual outliers. These are samples which have the same ancestry as bulk of the samples, however still show somewhat different genetic makeup. This might be due to considerable population stratification (different populations or ethnicities in the same superpopulation) or due to technical reasons. By default, we remove clear outliers.

#### Outlier samples

On the next plot you can see the distribution of Probabilistic Local Outlier Factor "outlierness" statistic ([Kriegel et al., 2009](https://dl.acm.org/doi/10.1145/1645953.1646195), [Privé et al., 2020](https://academic.oup.com/bioinformatics/article/36/16/4449/5838185)), based on which the samples are declared genetic outliers and removed from the analysis. Use the histogram to check if the default value is appropriate. If there is a long tail to the right, please adjust the default value in the pipeline argument `--GenOutThresh` and re-run the pipeline. Threshold should be selected so that there are no dramatic outliers on the first panes of *Fig 9*.

If you observe bimodal distribution on the histogram then it might indicate substratification in the genotype data. Check the following section, *Fig 5* and *Fig 6*. 

![**Fig 7. Statistic of outlierness with threshold used to remove ethnic outliers.**](outputfolder_gen/gen_plots/PC_dist_outliers_S.png)  

On the next two plots we visualise genetic principal components to check if genetic outliers were removed. If you observe unsolved sub-clusters on following PCA plots, this might indicate that you have different genotyping batches in the data, or large fraction of samples (e.g. >10%) from somewhat different ancestry. In this case, the algorithm might not assign all deviating samples as outliers or remove unreasonably large fraction of samples. 

If the pipline removes unreasonably large fraction of the samples, there are multiple courses of action to consider:

1. If you observe clusters with limited numbers of samples (e.g. N<500) and a clear separation based on PC1 or PC2, you can consider using the data as it is: first genetic PCs will be used as covariates in the meta-analysis and such batch effects will be mostly accounted for. In that case, you should re-run the `DataQC` pipeline with large SD value, e.g. `--GenSdThresh 10`, to keep all the data in.

2. If you have multiple batches with many samples (N>500), you might consider splitting the data based on genotyping batch and re-running `DataQC` and all following pipelines on those batches.

3. If splitting the data by batches is unoptimal (e.g. leads to removal of unreasonably many samples), you could add the batch as covariate into separate table and include it as `--AdditionalCovariates` argument of `DataQC`.

If the pipelines does not remove all clear outlier samples, adjust `--GenSdThresh` threshold accordingly, to remove small fraction of clearly deviating samples.

In case of doubt please contact the lead analyst(s) for advice and we will investigate further.

![**Fig 8. PCA with all samples, outliers outlined with red.** Different shades of red indicate different classes of outliers: difference in SD from mean of PC1 and PC2 (pink), statistic of outlierness S (red), or both (dark red). Dashed lines indicate the SD thresholds to consider samples to outliers.](outputfolder_gen/gen_plots/PCA_outliers.png)  

![**Fig 9. PCA with all samples, outliers removed.** You should not see any dramatic outliers or clusters on the first few PCs any more. These are the PCs which will be used as covariates in the eQTL meta-analysis.](outputfolder_gen/gen_plots/Target_PCs_postQC.png)

If you still observe clusters in the data, that could indicate that you still have different genotype batches in the data. If you know the genotype batches and splitting the data is unoptimal, you should specify the batch as additional covariate, using `--AdditionalCovariates` argument.

Next we visualise on the scree plot which genotype PCs are the most informative. This helps to make an informed decision, how many genotype PCs should be added to the association model as covariates.

![**Fig 10. Scree plot for PCA with all samples, outliers removed.** Visualised are the first 10 singular values.](outputfolder_gen/gen_plots/Target_PCs_scree_postQC.png)

#### Loadings of PCs

Visualise the loadings of 10 first PCs calculated on final quality-controlled genotype data. Because PC calculation command removed long-range LD, you should not see any clear peaks on the plots any more, those would indicate that PC is still driven variants in long-range LD regions.

![**Fig 11. Loadings of 10 first PCs.**](outputfolder_gen/gen_plots/Target_PCs_postQC_Loadings.png)

## Gene expression QC report

### Overview of the QC steps

```{r, message = FALSE, warning = FALSE, echo = FALSE}
overview_table <- fread("outputfolder_exp/exp_data_summary/summary_table.txt")
knitr::kable(overview_table)
```

### PCA on preprocessed expression data

On the following plots, 8 first PCs are calculated on normalised and log~2~-transformed expression matrix. Samples were declared as outliers based on first two PCs, if they deviate from the mean value by `r params$SD_exp` standard deviations (SDs). Analysis was run iteratively, data was pre-processed and PCA was calculated until there were no further expression otliers to remove. Because there might be expected clustering based on the sex, if possible, we also indicate the genetic/reported sex of each sample on the plot.

![**Fig 12. PCA on all samples** Outliers from all iterations are outlined as red. If available, genetic/reported sex of the individual is shown as the shape of the dot.](outputfolder_exp/exp_plots/PCA_before.png)

In order to visualise non-linear patterns in the data, we calculated 2 first multidimensional scaling (MDS) components and visualise those as well. This analysis were done on the data where outliers from previous step were already removed. Similarly to PCs, samples were declared as outliers when they deviate from the mean value of MDS 1 and MDS 2 by `r params$SD_exp` SDs. If you still observe outlier groups on this plot, you should try to reduce the SDs allowed to deviate (parameter `--ExpSdThreshold` of the pipeline) and re-run the pipeline.

![**Fig 13. MDS on all samples** Genetic/reported sex of the individual is shown as the shape of the dot, outliers are outlined as red.](outputfolder_exp/exp_plots/MDS_before.png)

These are 8 first expression PCs, calculated on fully processed expression data (normalised and INT-transformed) after removal of outlier samples. Genetic sex of the individual is shown as shape of the dot. You should not see any dramatic outliers any more on the first two PCs.

![**Fig 14. PCA on all samples, outliers removed.** Genetic/reported sex of the individual is shown as the shape of the dot.](outputfolder_exp/exp_plots/PCA_after.png) 

#### Explained variance of expression PCs

![**Fig 15. PCA scree plot for normalised and inverse normal transformed expression matrix.** Outlier samples were removed before re-calculating PCA.](outputfolder_exp/exp_plots/PCA_final_scree_plot.png)

Explained variance of final 100 first PCs which are used as covariates in the analysis (normalised and inverse normal transformed expression matrix).

```{r, message = FALSE, warning = FALSE, fig.width = 14, fig.height = 4, fig.align = "center"}
overview_pcs <- fread("outputfolder_exp/exp_data_summary/summary_pcs.txt")

overview_pcs_cumulative <- overview_pcs

overview_pcs_cumulative$cumulative <- cumsum(overview_pcs$explained_variance)

overview_pcs_cumulative$PC <- factor(overview_pcs_cumulative$PC, levels = as.character(overview_pcs_cumulative$PC))

p <- ggplot(overview_pcs_cumulative, aes(x = PC, y = cumulative)) + 
geom_bar(stat = "identity") + 
theme_bw() + 
scale_y_continuous(limits = c(0, 1)) +
theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

p
```

**Fig 16. Cumulative explained variance for first 100 PCs.**

First 5 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[5] * 100`% of variance.

First 10 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[10] * 100`% of variance.

First 20 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[20] * 100`% of variance.

First 50 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[50] * 100`% of variance.

First 100 PCs explain cumulatively `r overview_pcs_cumulative$cumulative[100] * 100`% of variance.

```{r, message = FALSE, warning = FALSE}
knitr::kable(overview_pcs)
```

### Sex-based mix-up check

`r if (!file.exists("outputfolder_exp/exp_plots/SexSpecificGenesXIST.png")) {"\\begin{comment}"}`

The next plot shows the expression of the sex chromosomes genes. We assume that males have higher expression of genes encoded from Y chromosome and females have higher expression of X chromosome gene *XIST*. Therefore, you should see that subsets of samples align parallel to x-axis (females) and y-axis (males). There should be a clear separation between males and females: samples in between the two clusters are likely samples that are contaminated.
We have drawn a line `r if(params$ContRatio == "fixed"){"corresponding to the x-y diagonal"} else {"from the origin to the respective maximum values of the x and y axes"}`. If you observe clear separation between males and females, and that the line does not capture this separation well, you may adjust the Slope of the line using the
`--ContaminationSlope` parameter. The slope is defined by a degree from 0 to 90. 45 degrees is the default. 0 degrees corresponds to a horizontal line, while 90 degrees corresponds to a vertical line.
Samples in the shaded area are automatically removed. The area is defined by an angle of `r params$Cont` degrees centered over the x-y diagonal. This is reasonable default, however you can adjust this with `--ContaminationArea` flag, if needed. The ratio between x and y axes have been set to be `r params$ContRatio`. This can be changed with `--XyExpressionRatio`.

Samples for which the sex-specific expression does not align with genetic sex are potential sample mix-ups.
These are outlined with red and automatically removed from further analyses. If the X chromosome was not present in the genotype data, expression is compared to reported sex instead.

![**Fig 17. Expression of X/Y chromosome genes.** On the x-axis is the normalised expression level of *XIST*, on the y-axis is the mean of the normalised expression levels of several Y chromosome genes. For the better separation of groups, minimal X and Y value was extracted from each value on X and Y axes. Point type indicates the genetic sex and there should be no samples on the X-Y diagonal. Samples which are in the wrong sample cloud, are likely sample mixups and were automatically removed from the data. Samples which are in the shaded area, are likely contaminated and automatically removed. Start of the separating line is set at median Y value of female samples and median X value of male samples.](`r ifelse(params$ContRatio == "fixed", "outputfolder_exp/exp_plots/SexSpecificGenesXIST.png", "outputfolder_exp/exp_plots/SexSpecificGenesXIST_zoomed.png")`)

`r if (!file.exists("outputfolder_exp/exp_plots/SexSpecificGenesXIST.png")) {"\\end{comment}"}`

`r if (!file.exists("outputfolder_exp/exp_plots/SexSpecificGenes.png")) {"\\begin{comment}"}`

The next plot shows the expression of the sex chromosomes genes. We assume that males have higher expression of genes encoded from Y chromosome and females have higher expression of X chromosome genes. Therefore, you should see that subsets of samples align parallel to x-axis (females) and y-axis (males). There should be a clear separation between males and females: samples in between the two clusters are likely samples that are contaminated.
We have drawn a line `r if(params$ContRatio == "fixed"){"corresponding to the x-y diagonal"} else {"from the origin to the respective maximum values of the x and y axes"}`. If you observe clear separation between males and females, and that the line does not capture this separation well, you may adjust the Slope of the line using the
`--ContaminationSlope` parameter. The slope is defined by a degree from 0 to 90. 45 degrees is the default. 0 degrees corresponds to a horizontal line, while 90 degrees corresponds to a vertical line.
Samples in the shaded area are automatically removed. The area is defined by an angle of `r params$Cont` degrees centered over the x-y diagonal. This is reasonable default, however you can adjust this with `--ContaminationArea` flag, if needed. The ratio between x and y axes have been set to be `r params$ContRatio`. This can be changed with `--XyExpressionRatio`.

Samples for which the sex-specific expression does not align with genetic sex are potential sample mix-ups.
These are outlined with red and automatically removed from further analyses. If the X chromosome was not present in the genotype data, expression is compared to reported sex instead.

![**Fig 18. Expression of X/Y chromosome genes.** On the x-axis is the mean normalised expression level of X chromosome genes, on the y-axis is the mean of the normalised expression levels of Y chromosome genes. For the better separation of groups, minimal X and Y value was extracted from each value on X and Y axes. Point type indicates the genetic sex and there should be no samples on the blue line (X-Y diagonal by default). Samples which are in the wrong sample cloud, are likely sample mixups and were automatically removed from the data. Samples which are in the shaded area, are likely contaminated and automatically removed. Start of the separating line is set at median Y value of female samples and median X value of male samples.](outputfolder_exp/exp_plots/SexSpecificGenes.png)

`r if (!file.exists("outputfolder_exp/exp_plots/SexSpecificGenes.png")) {"\\end{comment}"}`

`r if (file.exists("outputfolder_exp/exp_plots/SexSpecificGenes.png") || file.exists("outputfolder_exp/exp_plots/SexSpecificGenesXIST.png")) {"\\begin{comment}"}`

Warning! This step was skipped since relevant gene expression data from X and/or Y chromosome(s) was not available!

If possible, use the expression data which includes sex chromosomes.

**Fig 18. Expression of X/Y chromosome genes.** Not available!

`r if (file.exists("outputfolder_exp/exp_plots/SexSpecificGenes.png") || file.exists("outputfolder_exp/exp_plots/SexSpecificGenesXIST.png")) {"\\end{comment}"}`

# Acknowledgements

This report was generated with the help of following bioinformatics packages:

`bigsnpr`

URL: https://privefl.github.io/bigsnpr/

**Citations:**

<div class="csl-entry">Privé, F., Luu, K., Blum, M. G. B., McGrath, J. J., &#38; Vilhjálmsson, B. J. (2020). Efficient toolkit implementing best practices for principal component analysis of population genetic data. <i>Bioinformatics</i>, <i>36</i>(16), 4449–4457. https://doi.org/10.1093/BIOINFORMATICS/BTAA520</div>

\newline

<div class="csl-entry">Prive, F., Aschard, H., Ziyatdinov, A., &#38; Blum, M. G. B. (2018). Efficient analysis of large-scale genome-wide data with two R packages: bigstatsr and bigsnpr. <i>Bioinformatics</i>, <i>34</i>(16), 2781–2787. https://doi.org/10.1093/BIOINFORMATICS/BTY185</div>

\newline

`plink 1.9`

URL: https://www.cog-genomics.org/plink/

`plink 2`

URL: https://www.cog-genomics.org/plink/2.0/

**Citations:**

<div class="csl-entry">Purcell, S., Neale, B., Todd-Brown, K., Thomas, L., Ferreira, M. A. R., Bender, D., Maller, J., Sklar, P., de Bakker, P. I. W., Daly, M. J., &#38; Sham, P. C. (2007). PLINK: A tool set for whole-genome association and population-based linkage analyses. <i>American Journal of Human Genetics</i>, <i>81</i>(3), 559–575. https://doi.org/10.1086/519795</div>

\newline

<div class="csl-entry">Chang, C. C., Chow, C. C., Tellier, L. C. A. M., Vattikuti, S., Purcell, S. M., &#38; Lee, J. J. (2015). Second-generation PLINK: Rising to the challenge of larger and richer datasets. <i>GigaScience</i>, <i>4</i>(1). https://doi.org/10.1186/s13742-015-0047-8</div>

\newline

`edgeR`

URL: https://bioconductor.org/packages/release/bioc/html/edgeR.html

**Citations:**

<div class="csl-entry">Robinson, M. D., McCarthy, D. J., &#38; Smyth, G. K. (2010). edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. <i>Bioinformatics</i>, <i>26</i>(1), 139–140. https://doi.org/10.1093/BIOINFORMATICS/BTP616</div>

\newline

`PreprocessCore`

URL: https://bioconductor.org/packages/release/bioc/html/preprocessCore.html

**Citations:**

<div class="csl-entry">Bolstad, B. (2022). <i>preprocessCore: A collection of pre-processing functions</i>. https://github.com/bmbolstad/preprocessCore</div>
